File: Source/C#/DevnologyFitnesseDojo/DevnologyFitnesseDojo/Domain/Inventory.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DevnologyFitnesseDojo.Domain
{
    public class Inventory
    {
        private static readonly Inventory inventory = new Inventory();

        private readonly Dictionary<Book, int> books = new Dictionary<Book, int>();
        private readonly IList<PromoPackage> promoPackages = new List<PromoPackage>();

        public static Inventory Instance
        {
            get { return inventory; }
        }

        public void AddBook(Book book, int amount)
        {
            if (!books.ContainsKey(book))
            {
                books.Add(book, 0);
            }

            int currentAmount = books[book];

            currentAmount += amount;

            books[book] = currentAmount;
        }

        public Book FindByTitle(string title)
        {
            return books.Keys.SingleOrDefault(book => book.Title == title);
        }

        public Book FindByIsbn(string isbn)
        {
            return books.Keys.SingleOrDefault(book => book.Isbn == isbn);
        }

        public int CountBooks(Book book)
        {
            if (!books.ContainsKey(book))
            {
                return 0;
            }

            return books[book];
        }

        public void DeductInventory(Book book, int amount)
        {
            int stock = books[book];

            if (stock - amount < 0)
            {
                throw new ArgumentException("Not enough books in stock: current stock is " + stock + ", trying to deduct " +
                    amount);
            }

            stock -= amount;

            books[book] = stock;
        }

        public void AddPromoPackage(PromoPackage promoPackage)
        {
            promoPackages.Add(promoPackage);
        }


        public PromoPackage GetPromoPackage(Order order)
        {
            foreach (PromoPackage promoPackage in promoPackages)
            {
                if (promoPackage.IsEligible(order))
                {
                    return promoPackage;
                }
            }
            return null;
        }

        public bool IsPromoOrder(Order order)
        {
            foreach (PromoPackage promoPackage in promoPackages)
            {
                if (promoPackage.IsEligible(order))
                {
                    return true;
                }
            }
            return false;
        }
    }
}

File: Source/Java/nl/devnology/domain/book/Inventory.java
package nl.devnology.domain.book;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import nl.devnology.domain.order.Order;

/**
 * @author Erik Pragt
 */
public class Inventory {
    private static Inventory inventory = new Inventory();

    private static Map<Book, Integer> books = new HashMap<Book, Integer>();
    private static List<PromoPackage> promoPackages = new ArrayList<PromoPackage>();

    public static Inventory get() {
        return inventory;
    }


    public void addBook(Book book, Integer amount) {
        Integer currentAmount = books.get(book);

        if(currentAmount == null) {
            currentAmount = amount;
        } else {
            currentAmount += amount;
        }

        books.put(book, currentAmount);
    }

    public Book findByTitle(String title) {
        for(Map.Entry<Book, Integer> entry: books.entrySet()) {
            Book book = entry.getKey();
            if(book.getTitle().equals(title)) {
                return book;
            }
        }

        return null;
    }

    public Book findByIsbn(String isbn) {
        for(Map.Entry<Book, Integer> entry: books.entrySet()) {
            Book book = entry.getKey();
            if(book.getIsbn().equals(isbn)) {
                return book;
            }
        }

        return null;

    }

    public Integer countBooks(Book book) {
        return books.get(book);
    }

    public void deductInventory(Book book, Integer amount) {
        Integer stock = books.get(book);

        if(stock - amount < 0) {
            throw new IllegalArgumentException("Not enough books in stock: current stock is " + stock + ", trying to deduct " + amount);
        }

        stock -= amount;

        books.put(book, stock);
    }

    public void addPromoPackage(PromoPackage promoPackage) {
        promoPackages.add(promoPackage);
    }


    public PromoPackage getPromoPackage(Order order) {
        for (PromoPackage promoPackage : promoPackages) {
            if(promoPackage.isEligible(order)) {
                return promoPackage;
            }
        }
        return null;

    }

    public boolean isPromoOrder(Order order) {
        for (PromoPackage promoPackage : promoPackages) {
            if(promoPackage.isEligible(order)) {
                return true;
            }
        }
        return false;
    }
}
